[cxx-interop] A few cleanups and fixes for function templates.

zoecarver opened the pull request:
1. Correctly map `T&` -> `inout T`.
2. Fix issue with templates that return a dependent type.
3. Correctly represent function template names. IDE/debugger will print out `test` not `_ZtestSomeMangledThing`.
4. Remove a lot of code.

zoecarver committed:
[cxx-interop] Remove `generateSpecializedCXXFunctionTemplate` and just import the decl directly.

1) This unifies two code paths that do essentially the same thing.
2) This fixes how we import certain types (such as return types).

zoecarver committed:
[cxx-interop] Correctly map `T&` -> `inout T`.

This is important because the specialized template will be imported as `inout int` (for example) and we must make sure to match the signature.

Updates tests based on this and fixes a few tests that started failing after daceecfc7595d3cba32b36469e2d3db95491a1fe.

zoecarver commented on test/Interop/Cxx/templates/function-template.swift:
```diff
@@ -28,15 +28,6 @@ FunctionTemplateTestSuite.test("lvalueReference<T> where T == Int") {
   expectEqual(value, 42)
 }
 
-// TODO: currently "Any" is imported as an Objective-C "id".
```
This might look like a regression, but this should never have really been tested. It just happens to work. It was never designed, implemented, or meant to be supported.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test macOS.

zoecarver commented:
@swift-ci please test Windows

zoecarver commented:
@swift-ci Please Build Toolchain macOS Platform

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

hyp commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -395,18 +395,21 @@ ClangTypeConverter::reverseBuiltinTypeMapping(StructType *type) {
     // imported as an Int32 and CLongLong as an Int64. Therefore, manually
     // add mappings to C for Int and UInt.
     // On 64-bit Cygwin, no manual mapping is required.
-    if (Triple.isOSWindows() && Triple.isArch64Bit()
```
Can you delete this code instead of commenting it out?

hyp commented on test/Interop/Cxx/templates/function-template-irgen-objc.swift:
```diff
@@ -1,17 +0,0 @@
-// RUN: %target-swift-emit-ir %s -I %S/Inputs -enable-cxx-interop | %FileCheck %s
-
-// TODO: This needs to be fixed in the SwiftTypeConverter.
```
Why is this test deleted?

zoecarver replied:
Same reason as the test was removed from the other test file: 
> This might look like a regression, but this should never have really been tested. It just happens to work. It was never designed, implemented, or meant to be supported.

hyp commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -4672,8 +4672,7 @@ clang::FunctionDecl *ClangImporter::instantiateCXXFunctionTemplate(
                        {func->getName(), StringRef(failedTypesStr)});
-    // Return a valid FunctionDecl but, we'll never use it.
-    return func->getAsFunction();
+    return nullptr;
```
Is there a test case that tests this change? I couldn't spot one in your test cases.

zoecarver replied:
This is just moved up. There's already a test that covers this (where we verify clang's substitution failure errors).

hyp commented on lib/SILGen/SILGenBridging.cpp:
```diff
@@ -718,6 +718,12 @@ static ManagedValue emitNativeToCBridgedNonoptionalValue(SILGenFunction &SGF,
   // Fall back to dynamic Any-to-id bridging.
   // The destination type should be AnyObject in this case.
+  if (!bridgedType->isEqual(SGF.getASTContext().getAnyObjectType())) {
+    llvm::errs() << "Fallback not viable. About to crash. \n";
```
would it be better to have a user-visible frontend error that prohibits this bridging?

hyp commented on lib/Sema/CSApply.cpp:
```diff
@@ -285,8 +226,14 @@ Solution::resolveConcreteDeclRef(ValueDecl *decl,
                 subst);
-    auto newDecl = generateSpecializedCXXFunctionTemplate(
-        decl->getASTContext(), cast<AbstractFunctionDecl>(decl), subst, newFn);
+    llvm::errs() << "NEW FN: "; newFn->dump();
```
Why are we dumping the function here? Please don't commit the dumping code.

hyp commented on lib/Sema/CSApply.cpp:
```diff
@@ -285,8 +226,14 @@ Solution::resolveConcreteDeclRef(ValueDecl *decl,
+    llvm::errs() << "NEW FN: "; newFn->dump();
+    // We failed to specialize this function template. The compiler is going to
+    // exit soon. Return something valid in the meantime.
+    if (!newFn)
```
Is this condition tested? Should there be an error presented to the user if the compiler is going to exit (is it going to crash or report an error)?

zoecarver replied:
This is the exact same thing I removed from `instantiateCXXFunctionTemplate` (that you commented on above). It's the exact same thing, just moved around.

zoecarver commented:
@swift-ci please smoke test.

zoecarver committed:
[cxx-interop] Fix (U)Int substituted params on Windows.

Re-write the parameter list to use (U)Int if that was one of the substituted parameters. This is required because Windows will incorrectly map Int -> long long -> Int64.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver commented:
@swift-ci please test Windows platform.

zoecarver committed:
[nfc] Clang format the PR.

zoecarver commented on lib/Sema/CSApply.cpp:
```diff
@@ -304,9 +252,71 @@ Solution::resolveConcreteDeclRef(ValueDecl *decl,
+      }
+    }
+
+    if (auto fn = dyn_cast<AbstractFunctionDecl>(newDecl)) {
```
@hyp do you think it would be good to factor this if/hack out into it's own function?

zoecarver committed:
Move up the Windows hack and fix it for static members/inits.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please test Windows

zoecarver commented:
@swift-ci please smoke test.

zoecarver committed:
[nfc] [cxx-interop] Enable method tests on windows.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test Linux.

shahmishal commented:
```
Failed Tests (6):
  Swift(watchsimulator-i386) :: Interop/Cxx/namespace/templates-second-header.swift
  Swift(watchsimulator-i386) :: Interop/Cxx/namespace/templates.swift
  Swift(watchsimulator-i386) :: Interop/Cxx/templates/defaulted-template-type-parameter.swift
  Swift(watchsimulator-i386) :: Interop/Cxx/templates/function-template.swift
  Swift(watchsimulator-i386) :: Interop/Cxx/templates/member-templates.swift
  Swift(watchsimulator-i386) :: Interop/Cxx/templates/template-type-parameter-not-in-signature.swift
```
This introduced watchOS failures.
